### PR TITLE
Artifact filtering

### PR DESCRIPTION
**Pull request by xMAC94x**

### CODE REVIEW

File: server/src/models/artifact.rs [hunk 1]
```diff
@@ -48,7 +48,12 @@ impl Artifact {
             .expect("Failed to parse date!");
             let build_id = build.id as i32;
             let platform = Self::get_platform(&build.name)?;
-            let channel = Self::get_channel();
+            let channel = if attributes.variables.iter().any(|e| e.key == "nightly") {
```
Comment (Songtronix) [inline server/src/models/artifact.rs:51]: We probably want to expose this as a setting and log that we ignored that event as it didn't match the variable filter. So rather implement as a check in https://github.com/Songtronix/Airshipper/blob/master/server/src/models/gitlab.rs#L20

File: server/src/models/gitlab.rs [hunk 2]
```diff
@@ -55,7 +61,7 @@ pub struct ObjectAttributes {
     pub created_at: Option<String>,
     pub finished_at: Option<String>,
     pub duration: Option<u64>,
-    pub variables: Vec<String>,
+    pub variables: Vec<Variable>,
```
Comment (Songtronix) [inline server/src/models/gitlab.rs:64]: Has this been tested?

**Review by Songtronix — CHANGES_REQUESTED** [review]

File: server/src/models/gitlab.rs (hunk 2 above)
Comment (xMAC94x) [reply server/src/models/gitlab.rs:64]: yes

File: server/.env.example [hunk 3]
```diff
@@ -4,6 +4,7 @@ DATABASE_URL=airshipper.db
 AIRSHIPPER_GITLAB_SECRET=banana
 AIRSHIPPER_ARTIFACT_STAGE=post
 AIRSHIPPER_TARGET_BRANCH=master
+#AIRSHIPPER_TARGET_VARIABLE=nightly
```
Comment (Songtronix) [inline server/.env.example:7]: not needed to comment that out. It's an example anyway to showcase possible values for these.

File: server/src/config/mod.rs [hunk 4]
```diff
@@ -49,6 +53,8 @@ impl ServerConfig {
             spaces_cdn: Self::expect_env_key("AIRSHIPPER_SPACES_CDN").parse().unwrap_or(true),
             // Optional
             target_branch: Self::get_env_key_or("AIRSHIPPER_TARGET_BRANCH", "master"),
+            // Optional
```
Comment (Songtronix) [inline server/src/config/mod.rs:56]: Not needed to duplicate the comment imo

File: server/src/config/mod.rs [hunk 5]
```diff
@@ -23,6 +23,10 @@ pub struct ServerConfig {
     pub artifact_stage: String,
     /// What branch should be downloaded
     pub target_branch: String,
+    /// What variable should be expected to be present in the Pipeline Webhook from gitlab.
+    /// NOTE: an empty `target_variable` will pass all artifacts
+    /// NOTE: if both `target_branch` and `target_variable` is present, both filters will be applied
+    pub target_variable: String,
```
Comment (Songtronix) [inline server/src/config/mod.rs:29]: probably good to make it an Option which clarifies and encodes the intent into the typesystem.

I would rephrase the comment to:
```
/// If set only pipeline events with this variable will be processed. Otherwise every event will be considered.
```

File: server/src/models/gitlab.rs [hunk 6]
```diff
@@ -26,6 +26,17 @@ impl PipelineUpdate {
...
         }
 
+        if !CONFIG.target_variable.is_empty()
+            && !self
+                .object_attributes
+                .variables
+                .iter()
+                .any(|e| e.key == CONFIG.target_variable)
+        {
+            tracing::debug!("Variable was not found '{}'", CONFIG.target_variable);
```
Comment (Songtronix) [inline server/src/models/gitlab.rs:36]: Be explicit what was the expected variable to be found like https://github.com/Songtronix/Airshipper/blob/7ba15179df49063c3fb6b3976b061e7dcd3516e3/server/src/models/gitlab.rs#L22

**Review by Songtronix — CHANGES_REQUESTED** [review]

File: server/src/models/gitlab.rs (hunk 6 above)
Comment (xMAC94x) [reply server/src/models/gitlab.rs:36]: i am already listing the expected variable to be found. do you want to to list the vec of all provided variables, even if it can be very large?
